oauthproxy.go

```diff
@@ -135,38 +135,144 @@ func (u *UpstreamProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+//
+//    But the default path appending behavior of httputil's version would cause
+//    extra path info being injected (e.g. "http://upstream/upath/upath/file.html")
+//    resulting in incorrect upstream URLs. Further, we'd like to remap some
+//    upstream locations by explicitly providing a different "base", for example:
+//
+//         "http://proxy/file.html" => "http://upstream/upath/file.html"
+//
+//    Can be accomplished by specifying "http://upstream/upath/#/" and it will
+//    use the fragment as the `muxPath` (similar to how local file remapping works).
+//
+//  - Also, there (at the time of writing) is a bug with how it handles escaped Paths:
+//      https://github.com/golang/go/pull/36378/files
+//
+//  - However, we make sure to retain the query param merging, as it's important
+//    in situations where the upstream requires an authentication token for example.
+//
+func NewReverseProxy(target *url.URL, muxPath string, opts *Options) (proxy *httputil.ReverseProxy) {
+
+	passRequestHost := opts.PassHostHeader
 	proxy = httputil.NewSingleHostReverseProxy(target)
+
+	primeDirector := proxy.Director
+
+	proxy.Director = func(req *http.Request) {
+		// Save the original path (and rawpath) to work around the bug with
+		// escaped slashes in older versions of Go.
+		origPath, origRawPath := req.URL.Path, req.URL.RawPath
+
+		// Wrap the default httputil implementation as it does some useful
+		// things like query string merging and updating the User Agent if
+		// necessary (also lets us get fixes made to Go later on).
+		primeDirector(req)
+
+		if !passRequestHost {
+			// Make sure this matches the target if we're not explicitly
+			// sending the request's Host instead.
+			req.Host = target.Host
+		}
+
+		// Take request's "/foo/file.html" and strip the muxPath off the
+		// front. We know it exists on the path since that's how this
+		// handler got called in the first place (via proxy's serveMux
+		// path matching). If muxPath is "/foo/" we want "/file.html"
+		// (note the slash is retained). Note: this replaces the possibly
+		// incorrect path calculated in the default implementation.
+
+		if origRawPath != "" {
+			// If this is not empty, then the path had some escaping done,
+			// and in order to maintain its pairity with the "non-Raw"
+			// version, we have to trim that one so it matches. As the
+			// semantics are that if RawPath doesn't decode into exactly
+			// the Path value, it's ignored.
+			req.URL.RawPath = strings.TrimPrefix(origRawPath, muxPath)
+			if unescapedMuxPath, err := url.PathUnescape(muxPath); err != nil {
+				req.URL.Path = strings.TrimPrefix(origPath, unescapedMuxPath)
+			} else {
+				// This would only happen if muxPath was incorrectly passed
+				// to us (such a malformed manual path remapping). Not really
+				// much we can do... I'm not sure what HTTP servers are
+				// supposed to do in this case.
+				req.URL.Path = strings.TrimPrefix(origPath, muxPath)
```

Comment:
Would it make sense to error out?

Reply:
You could toss a 500 error here I suppose...

Reply:
I think a 500 would make sense here

Reply:
Something didn't sit right with me about this logic. I've taken another pass that is much cleaner and more robust. A 500 here is unnecessary now.